bundles/model/org.eclipse.smarthome.model.script/src/org/eclipse/smarthome/model/script/lib/NumberExtensions.java

```diff
@@ -219,6 +223,111 @@ public static boolean operator_lessEqualsThan(Type type, Number x) {
         }
     }
 
+    // QuantityType support
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public static QuantityType<?> operator_plus(QuantityType<?> x, QuantityType<?> y) {
+        return x == null ? y : y == null ? x : x.add((QuantityType) y);
+    }
+
+    public static QuantityType<?> operator_minus(QuantityType<?> x) {
+        return x == null ? null : x.negate();
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public static QuantityType<?> operator_minus(QuantityType<?> x, QuantityType<?> y) {
+        return x == null ? operator_minus(y) : y == null ? x : x.subtract((QuantityType) y);
+    }
+
+    public static QuantityType<?> operator_multiply(Number x, QuantityType<?> y) {
+        BigDecimal xValue = numberToBigDecimal(x);
+        if (xValue == null) {
+            return new QuantityType<>("0");
+        } else if (y == null) {
+            return new QuantityType<>("0");
+        } else {
+            return y.multiply(xValue);
+        }
+    }
+
+    public static QuantityType<?> operator_multiply(QuantityType<?> x, Number y) {
+        return operator_multiply(y, x);
+    }
+
+    public static QuantityType<?> operator_multiply(QuantityType<?> x, QuantityType<?> y) {
+        return x == null || y == null ? new QuantityType<>("0") : x.multiply(y);
+    }
+
+    public static QuantityType<?> operator_divide(QuantityType<?> x, Number y) {
+        BigDecimal yValue = numberToBigDecimal(y);
+        if (x == null) {
+            return new QuantityType<>("0").divide(yValue);

```

Comment:
what would you expect when you divide nothing by anything? Its just for the sake of completeness I guess. we can as well return QuantityType.ZERO (dimensionless) instead. See https://github.com/eclipse/smarthome/blob/master/bundles/model/org.eclipse.smarthome.model.script/src/org/eclipse/smarthome/model/script/lib/NumberExtensions.java#L85 for comparison.

Reply:
> what would you expect when you divide nothing by anything?

Clearly ZERO, so there is no point in doing any calculations here.
But why do you at all say it is nothing? I would expect that "null" is treated as the neutral element of the operation. Just like you allow "+ 3" to be "0 + 3", I would expect a "/ 3" to be "1 / 3" - you divide "all there is, whatever it is" by three.

> See ... for comparison.

Yeah, nice bug ;-)
Who the hack has written that bad code? 😇

Reply:
> I would expect that "null" is treated as the neutral element of the operation

Is `1` really the neutral element here? For "0 + 3" its neutral because the result does not change. But "1 / 3" does change the result. If we want `null` to be treated as neutral we should just return the 2nd operand and also return the 1st operand in case the 2nd is `null`.
wdyt?

Reply:
and if we do changes here: is it API breaking? Number and QuantityType calculation should be aligned so changing the behaviour would possibly break existing rules.

Reply:
Maybe operations with one operand being `null` are invalid in the first place and we should return UnDefType.NULL?

Reply:
Definitely not UnDefType.NULL, but we could simply return null. What is Xbase doing for normal numbers in those cases?

Reply:
A demo rule with execution script `var test = null / 3` does throw an error. So it seems not to be supported.

Reply:
I rather thought about looking at their code in order to do the same :-)

Reply:
XBase definition just reflects that: multiply operations (which includes divide) is defined as `a <operation> b`. The above test script already gives a warning during parsing as not being valid.
My suggestion is to return null and let the script fail happily. It will be up to the user to do proper null checks before calling these operations.
The other question is: this will become API breaking when we align DecimalType & QuantityType arithmetic behaviour.

Reply:
"does throw an error" and "returns null" are imho two different behaviors, aren't they? So how does the code exactly look like in Xbase? Regarding breaking the API for DecimalType operations: That should be imho ok as it is anyhow only changing one error situation by another.

Reply:
```
public static BigDecimal operator_divide(BigDecimal a, BigDecimal b) {
    return a.divide(b, MathContext.DECIMAL128);
}
```

Reply:
> changing one error situation by another

Right now we do not produce an error when `null / 3` occurs. For Numbers we return `0`.

Reply:
So to sum it up: In order to do it the same way as Xbase, you could simply completely remove the x==null check?

Reply:
yes. an update is on the way.